Kathryn Brusewitz

# Trees and Forests

In [174]:
# Numpy is a library for working with Arrays
import numpy as np
print("Numpy version:        %6.6s" % np.__version__)

# SciPy implements many different numerical algorithms
import scipy as sp
print("SciPy version:        %6.6s" % sp.__version__)
import scipy.stats as st

# Pandas makes working with data tables easier
import pandas as pd
print("Pandas version:       %6.6s" % pd.__version__)

# Module for plotting
import matplotlib 
print("Maplotlib version:    %6.6s" % matplotlib.__version__)
%matplotlib inline
# needed for inline plots in notebooks
import matplotlib.pyplot as plt  

# SciKit Learn implements several Machine Learning algorithms
import sklearn
print("Scikit-Learn version: %6.6s" % sklearn.__version__)
import statsmodels.formula.api as smf
# import statsmodels.api as sm
import sklearn.metrics as sm


from scipy.stats import linregress

from matplotlib.colors import ListedColormap
from sklearn import neighbors

import seaborn as sns
sns.set(style="ticks")

from sklearn.cluster import KMeans
from sklearn import linear_model
from sklearn.linear_model import LogisticRegression
from sklearn.cross_validation import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn import metrics
from sklearn.metrics import accuracy_score, precision_score, recall_score
from sklearn.metrics import mean_squared_error

from sklearn.cross_validation import KFold

from sklearn.linear_model import LinearRegression

from sklearn.metrics import confusion_matrix

import math

Numpy version:        1.13.3
SciPy version:        0.19.1
Pandas version:       0.20.3
Maplotlib version:     2.1.0
Scikit-Learn version: 0.19.1


# 1 Explore the Data

## 1.1

In [175]:
# Load in data
df = pd.read_csv('titanic.csv')

df

pclass  survived                                               name  \
0          1         1                      Allen, Miss. Elisabeth Walton   
1          1         1                     Allison, Master. Hudson Trevor   
2          1         0                       Allison, Miss. Helen Loraine   
3          1         0               Allison, Mr. Hudson Joshua Creighton   
4          1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5          1         1                                Anderson, Mr. Harry   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
7          1         0                             Andrews, Mr. Thomas Jr   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9          1         0                            Artagaveytia, Mr. Ramon   
10         1         0                             Astor, Col. John Jacob   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12         1         1                      Aubart, Mme. Leontine Pauline   
13         1         1                       Barber, Miss. Ellen "Nellie"   
14         1         1               Barkworth, Mr. Algernon Henry Wilson   
15         1         0                                Baumann, Mr. John D   
16         1         0                           Baxter, Mr. Quigg Edmond   
17         1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18         1         1                              Bazzani, Miss. Albina   
19         1         0                               Beattie, Mr. Thomson   
20         1         1                      Beckwith, Mr. Richard Leonard   
21         1         1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   
22         1         1                              Behr, Mr. Karl Howell   
23         1         1                              Bidois, Miss. Rosalie   
24         1         1                                  Bird, Miss. Ellen   
25         1         0                                Birnbaum, Mr. Jakob   
26         1         1                            Bishop, Mr. Dickinson H   
27         1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
28         1         1                             Bissette, Miss. Amelia   
29         1         1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   
...      ...       ...                                                ...   
1279       3         0               Vestrom, Miss. Hulda Amanda Adolfina   
1280       3         0                                    Vovk, Mr. Janko   
1281       3         0                               Waelens, Mr. Achille   
1282       3         0                                Ware, Mr. Frederick   
1283       3         0                        Warren, Mr. Charles William   
1284       3         0                                  Webber, Mr. James   
1285       3         0                                Wenzel, Mr. Linhart   
1286       3         1    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1287       3         0                   Widegren, Mr. Carl/Charles Peter   
1288       3         0                          Wiklund, Mr. Jakob Alfred   
1289       3         0                            Wiklund, Mr. Karl Johan   
1290       3         1                   Wilkes, Mrs. James (Ellen Needs)   
1291       3         0                   Willer, Mr. Aaron ("Abi Weller")   
1292       3         0                                 Willey, Mr. Edward   
1293       3         0                  Williams, Mr. Howard Hugh "Harry"   
1294       3         0                               Williams, Mr. Leslie   
1295       3         0                                Windelov, Mr. Einar   
1296       3         0                                   Wirz, Mr. Albert   
1297       3         0                             Wiseman, Mr. Phillippe   
1298       3         0                          Wittevrongel, Mr. Camille   
12

## 1.2
Create a summary table where you list the averages for those who survived and those who died.

Show means, ranges, and number of missings for each variable. In addition, add correlation between the price and the variable. You may add more statistics you consider useful to this table.

In [176]:
# get unique values for a column
for var in df.columns:
    print(var)
    print()
    values = df[var].unique()
    print(values[0:100])
    print('-------------------')

pclass

[1 2 3]
-------------------
survived

[1 0]
-------------------
name

['Allen, Miss. Elisabeth Walton' 'Allison, Master. Hudson Trevor'
 'Allison, Miss. Helen Loraine' 'Allison, Mr. Hudson Joshua Creighton'
 'Allison, Mrs. Hudson J C (Bessie Waldo Daniels)' 'Anderson, Mr. Harry'
 'Andrews, Miss. Kornelia Theodosia' 'Andrews, Mr. Thomas Jr'
 'Appleton, Mrs. Edward Dale (Charlotte Lamson)' 'Artagaveytia, Mr. Ramon'
 'Astor, Col. John Jacob'
 'Astor, Mrs. John Jacob (Madeleine Talmadge Force)'
 'Aubart, Mme. Leontine Pauline' 'Barber, Miss. Ellen "Nellie"'
 'Barkworth, Mr. Algernon Henry Wilson' 'Baumann, Mr. John D'
 'Baxter, Mr. Quigg Edmond'
 'Baxter, Mrs. James (Helene DeLaudeniere Chaput)' 'Bazzani, Miss. Albina'
 'Beattie, Mr. Thomson' 'Beckwith, Mr. Richard Leonard'
 'Beckwith, Mrs. Richard Leonard (Sallie Monypeny)' 'Behr, Mr. Karl Howell'
 'Bidois, Miss. Rosalie' 'Bird, Miss. Ellen' 'Birnbaum, Mr. Jakob'
 'Bishop, Mr. Dickinson H' 'Bishop, Mrs. Dickinson H (Helen Walton)'

In [178]:
# Init a temp table
temp = pd.DataFrame(data=df, columns = df.columns)

def sex_male(row):
    if(row['sex'] == 'male'):
        return 1
    else:
        return 0

def sex_female(row):
    if(row['sex'] == 'female'):
        return 1
    else:
        return 0

def embarked_S(row):
    if(row['embarked'] == 'S'):
        return 1
    else:
        return 0

def embarked_C(row):
    if(row['embarked'] == 'C'):
        return 1
    else:
        return 0

def embarked_Q(row):
    if(row['embarked'] == 'Q'):
        return 1
    else:
        return 0

# Apply functions to each row
temp['sex_female'] = temp.apply(lambda row: sex_female(row), axis=1)
temp['sex_male'] = temp.apply(lambda row: sex_male(row), axis=1)
temp['embarked_S'] = temp.apply(lambda row: embarked_S(row), axis=1)
temp['embarked_C'] = temp.apply(lambda row: embarked_C(row), axis=1)
temp['embarked_Q'] = temp.apply(lambda row: embarked_Q(row), axis=1)

# Init summary table with means of survived and drowned
summary = pd.DataFrame(temp.groupby('survived').mean()).T

summary.columns = ['survived', 'drowned']
categorical = temp.drop(['cabin', 'boat', 'home.dest', 'name', 'ticket'], axis = 1)

summary['null_count'] = categorical.isnull().sum()
summary['min'] = categorical.min()
summary['max'] = categorical.max()
summary['corr'] = ""

# Add correlation between fare and each variable
for i in ['pclass', 'age', 'sibsp', 'parch', 'fare', 'body', 'embarked_C', 'embarked_S', 'embarked_Q', 'sex_male', 'sex_female']:
    summary['corr'][i] = pd.Series.corr(temp[i], temp['fare'])

print(summary)

/Users/kathryn/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:54: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


              survived    drowned  null_count     min      max       corr
pclass        2.500618   1.962000           0       1        3  -0.558629
age          30.545369  28.918228         263  0.1667       80   0.178739
sibsp         0.521632   0.462000           0       0        8   0.160238
parch         0.328801   0.476000           0       0        9   0.221539
fare         23.353831  49.361184           1       0  512.329          1
body        160.809917        NaN        1188       1      328 -0.0431102
sex_female    0.156984   0.678000           0       0        1   0.185523
sex_male      0.843016   0.322000           0       0        1  -0.185523
embarked_S    0.754017   0.608000           0       0        1  -0.172683
embarked_C    0.148331   0.300000           0       0        1   0.286269
embarked_Q    0.097651   0.088000           0       0        1  -0.130059


# 2 Implement Decision Tree

Stay with binary-only variables. Split the data into just training and testing parts and avoid cross-validation.

## 2.1 Create binary variables

In [179]:
# 1. Create binary variables (dummies) for pclass: pc3, pc2, and pc1
def pclassBinary(row, expect):
    if(row['pclass'] == expect):
        return 1
    else:
        return 0

# Add binary variables for pclass
df['pc3'] = df.apply(lambda row: pclassBinary(row, 3), axis=1)
df['pc2'] = df.apply(lambda row: pclassBinary(row, 2), axis=1)
df['pc1'] = df.apply(lambda row: pclassBinary(row, 1), axis=1)

def ageBinary(row, minAge, maxAge):
    if(minAge <= row['age'] <= maxAge):
        return 1
    else:
        return 0

def oldAgeBinary(row, minAge):
    if(minAge <= row['age']):
        return 1
    else:
        return 0

# Add binary variables for age
df['age 0-10'] = df.apply(lambda row: ageBinary(row, 0, 10), axis=1)
df['age 11-17'] = df.apply(lambda row: ageBinary(row, 11, 17), axis=1)
df['age 18-45'] = df.apply(lambda row: ageBinary(row, 18, 45), axis=1)
df['age 46-59'] = df.apply(lambda row: ageBinary(row,46, 59), axis=1)
df['age 60-inf'] = df.apply(lambda row: oldAgeBinary(row, 60), axis=1)

# We already have binary variables for embarked C Q and S
# And we already have binary variables for female and male

print(df.columns)

Index(['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket',
       'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest', 'sex_female',
       'sex_male', 'embarked_S', 'embarked_C', 'embarked_Q', 'pc3', 'pc2',
       'pc1', 'age 0-10', 'age 11-17', 'age 18-45', 'age 46-59', 'age 60-inf'],
      dtype='object')


In [180]:
df

pclass  survived                                               name  \
0          1         1                      Allen, Miss. Elisabeth Walton   
1          1         1                     Allison, Master. Hudson Trevor   
2          1         0                       Allison, Miss. Helen Loraine   
3          1         0               Allison, Mr. Hudson Joshua Creighton   
4          1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5          1         1                                Anderson, Mr. Harry   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
7          1         0                             Andrews, Mr. Thomas Jr   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9          1         0                            Artagaveytia, Mr. Ramon   
10         1         0                             Astor, Col. John Jacob   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12         1         1                      Aubart, Mme. Leontine Pauline   
13         1         1                       Barber, Miss. Ellen "Nellie"   
14         1         1               Barkworth, Mr. Algernon Henry Wilson   
15         1         0                                Baumann, Mr. John D   
16         1         0                           Baxter, Mr. Quigg Edmond   
17         1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18         1         1                              Bazzani, Miss. Albina   
19         1         0                               Beattie, Mr. Thomson   
20         1         1                      Beckwith, Mr. Richard Leonard   
21         1         1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)   
22         1         1                              Behr, Mr. Karl Howell   
23         1         1                              Bidois, Miss. Rosalie   
24         1         1                                  Bird, Miss. Ellen   
25         1         0                                Birnbaum, Mr. Jakob   
26         1         1                            Bishop, Mr. Dickinson H   
27         1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
28         1         1                             Bissette, Miss. Amelia   
29         1         1          Bjornstrom-Steffansson, Mr. Mauritz Hakan   
...      ...       ...                                                ...   
1279       3         0               Vestrom, Miss. Hulda Amanda Adolfina   
1280       3         0                                    Vovk, Mr. Janko   
1281       3         0                               Waelens, Mr. Achille   
1282       3         0                                Ware, Mr. Frederick   
1283       3         0                        Warren, Mr. Charles William   
1284       3         0                                  Webber, Mr. James   
1285       3         0                                Wenzel, Mr. Linhart   
1286       3         1    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1287       3         0                   Widegren, Mr. Carl/Charles Peter   
1288       3         0                          Wiklund, Mr. Jakob Alfred   
1289       3         0                            Wiklund, Mr. Karl Johan   
1290       3         1                   Wilkes, Mrs. James (Ellen Needs)   
1291       3         0                   Willer, Mr. Aaron ("Abi Weller")   
1292       3         0                                 Willey, Mr. Edward   
1293       3         0                  Williams, Mr. Howard Hugh "Harry"   
1294       3         0                               Williams, Mr. Leslie   
1295       3         0                                Windelov, Mr. Einar   
1296       3         0                                   Wirz, Mr. Albert   
1297       3         0                             Wiseman, Mr. Phillippe   
1298       3         0                          Wittevrongel, Mr. Camille   
12

In [221]:
# Split data into training/testing groups (80/20% or so)
binaryVars = ['sex_male', 'embarked_S', 'embarked_C', 'embarked_Q', 'pc3', 'pc2',
       'pc1', 'age 0-10', 'age 11-17', 'age 18-45', 'age 46-59', 'age 60-inf']

x = df[binaryVars]
y = df['survived']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20, random_state=0)

## 2.2 Implement decision tree with binary variables

In [225]:
# Functions

# x = proportion survived
# y = proportion drowned

def ent(x, y):
    return -1 * x * math.log2(x) - y * math.log2(y)

# e0 = parent entropy
# e1, e2 = children entropies
# t0 = parent total samples
# t1, t2 = children total samples

def IG(e0, e1, e2, t0, t1, t2):
    return e0 - (((t1/t0)*e1) + ((t2/t0)*e2))

### 2.2.1
Compute the entropy at the root, based on the total number of survivors and drowned. You will use this number to compute entropy gain below

In [226]:
total = len(y_train)
pSurvived = (y_train == 1).sum()/total
pDrowned = (y_train == 0).sum()/total

rootEnt = ent(pSurvived, pDrowned)

print('Root Entropy: ', rootEnt) # 0.959472807639

Root Entropy:  0.959472807639


Root Entropy:  0.959472807639

### 2.2.2
Repeat the following with each feature you consider:

1. Split the data according to 0/1 answer to this feature. 
2. Compute entropy on both subsets
3. Compute the weighted average entropy 
4. Compute the entropy gain.

In [228]:
largestGain = 0
largestVar = ''

for var in binaryVars:
    
    # Split the data according to 0/1 answer to feature 'var'
    filter1 = (x_train[var] == 1)
    filter2 = (x_train[var] == 0)
    subset1 = y_train[filter1]
    subset2 = y_train[filter2]
    
    total = len(subset1)+len(subset2)
    
    # Compute entropy on both subsets
    subset1ProportionSurvived = (subset1 == 1).sum()/len(subset1)
    subset1ProportionDrowned = (subset1 == 0).sum()/len(subset1)
    subset2ProportionSurvived = (subset2 == 1).sum()/len(subset2)
    subset2ProportionDrowned = (subset2 == 0).sum()/len(subset2)
    
    subset1Entropy = ent(subset1ProportionSurvived, subset1ProportionDrowned)
    subset2Entropy = ent(subset2ProportionSurvived, subset2ProportionDrowned)
    
    print('subset 1 entropy: ', subset1Entropy)
    print('subset 2 entropy: ', subset2Entropy)
    
    # Compute entropy gain
    infoGain = IG(rootEnt, subset1Entropy, subset2Entropy, total, len(subset1), len(subset2))
    
    if (infoGain > largestGain):
        largestGain = infoGain
        largestVar = var
        
    print('info gain: ', infoGain)
    print('---')
    
print('--- Results ---')
print('Largest info gain: ', largestGain)
print('Variable: ', largestVar)

subset 1 entropy:  0.692005774194
subset 2 entropy:  0.853599095668
info gain:  0.208200714967
---
subset 1 entropy:  0.918743812538
subset 2 entropy:  1.0
info gain:  0.0171359855606
---
subset 1 entropy:  0.989245296929
subset 2 entropy:  0.923356553425
info gain:  0.0232154018563
---
subset 1 entropy:  0.9433656583
subset 2 entropy:  0.96096158207
info gain:  0.000141411437937
---
subset 1 entropy:  0.811278124459
subset 2 entropy:  0.995131720178
info gain:  0.0647845036508
---
subset 1 entropy:  0.992362874747
subset 2 entropy:  0.946712284891
info gain:  0.00342983866979
---
subset 1 entropy:  0.960230761497
subset 2 entropy:  0.886210771247
info gain:  0.0548100617444
---
subset 1 entropy:  0.998363672594
subset 2 entropy:  0.95292338153
info gain:  0.00381519656059
---
subset 1 entropy:  0.979868756651
subset 2 entropy:  0.958310230075
info gain:  0.000174221043957
---
subset 1 entropy:  0.969594464737
subset 2 entropy:  0.942759316574
info gain:  0.00105324700768
---
subset 1 

Largest info gain:  0.208200714967

Variable:  sex_male

### 2.2.3
Now pick the feature that gives the largest entropy gain. This gives you the  first split (tree stump). Print the survived/non-survived percentage in both branches.

In [229]:
filter1 = (x_train[largestVar] == 1)
filter2 = (x_train[largestVar] == 0)
subset1 = y_train[filter1]
subset2 = y_train[filter2]

subset1ProportionSurvived = (subset1 == 1).sum()/len(subset1)
subset1ProportionDrowned = (subset1 == 0).sum()/len(subset1)
subset2ProportionSurvived = (subset2 == 1).sum()/len(subset2)
subset2ProportionDrowned = (subset2 == 0).sum()/len(subset2)


print('subset 1 proportion survived: ', subset1ProportionSurvived)
print('subset 1 proportion drowned: ', subset1ProportionDrowned)
print('subset 2 proportion survived: ', subset2ProportionSurvived)
print('subset 2 proportion drowned: ', subset2ProportionDrowned)


subset 1 percent survived:  0.185520361991
subset 1 percent drowned:  0.814479638009
subset 2 percent survived:  0.721354166667
subset 2 percent drowned:  0.278645833333


subset 1 percent survived:  0.185520361991

subset 1 percent drowned:  0.814479638009

subset 2 percent survived:  0.721354166667

subset 2 percent drowned:  0.278645833333

Repeat the previous steps over and over again until you either

1. get a pure leaf
2. run out of data (number of observations in your leaf is too small, say <5)
3. run out of variables

